File: agentmail-invoice-processor/typescript/src/purchaseOrders.ts
/**
 * CSV-backed open purchase order list.
 *
 * `purchase_orders.csv` columns:
 *   po_number, vendor_name, amount, currency, description, status
 *
 * Match priority:
 *   1. Exact PO number (best signal — invoice cites the PO)
 *   2. Vendor name + amount within $1 (fallback for invoices missing PO ref)
 */

import * as fs from "node:fs";

const PO_FILE = "purchase_orders.csv";

export interface PurchaseOrder {
  po_number: string;
  vendor_name: string;
  amount: string; // CSV stores as string
  currency: string;
  description: string;
  status: string;
}

function parseCsvLine(line: string): string[] {
  const out: string[] = [];
  let cur = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (inQuotes) {
      if (c === '"' && line[i + 1] === '"') { cur += '"'; i++; }
      else if (c === '"') inQuotes = false;
      else cur += c;
    } else {
      if (c === '"') inQuotes = true;
      else if (c === ",") { out.push(cur); cur = ""; }
      else cur += c;
    }
  }
  out.push(cur);
  return out;
}

export function loadAll(): PurchaseOrder[] {
  if (!fs.existsSync(PO_FILE)) return [];
  const text = fs.readFileSync(PO_FILE, "utf8");
  const lines = text.split(/\r?\n/).filter((l) => l.length);
  if (!lines.length) return [];
  const headers = parseCsvLine(lines[0]);
  const rows: PurchaseOrder[] = [];
  for (let i = 1; i < lines.length; i++) {
    const cells = parseCsvLine(lines[i]);
    const row: any = {};
    headers.forEach((h, idx) => { row[h] = (cells[idx] ?? "").trim(); });
    rows.push(row as PurchaseOrder);
  }
  return rows;
}

export function findMatch(
  poNumber: string | null,
  vendorName: string | null,
  amount: number | null,
): PurchaseOrder | undefined {
  const rows = loadAll().filter(
    (r) => (r.status || "open").toLowerCase() === "open",
  );

  // Strategy 1: exact PO number match
  if (poNumber) {
    const poClean = poNumber.trim().toUpperCase();
    const m = rows.find((r) => (r.po_number || "").trim().toUpperCase() === poClean);
    if (m) return m;
  }

  // Strategy 2: vendor name + amount within $1
  if (vendorName && amount !== null && !isNaN(amount)) {
    const vClean = vendorName.trim().toLowerCase();
    for (const r of rows) {
      const vMatch = (r.vendor_name || "").trim().toLowerCase() === vClean;
      const rAmount = parseFloat(r.amount || "0");
      const aMatch = !isNaN(rAmount) && Math.abs(rAmount - amount) <= 1.0;
      if (vMatch && aMatch) return r;
    }
  }

  return undefined;
}

File: agentmail-invoice-processor/python/purchase_orders.py
"""
CSV-backed open purchase order list.

`purchase_orders.csv` columns:
  po_number, vendor_name, amount, currency, description, status

The agent matches incoming invoices against this list to decide auto-approve
vs escalate. Match priority:
  1. Exact PO number (best signal — invoice cites the PO)
  2. Vendor name + amount within $1 (fallback for invoices missing PO ref)
"""

import csv
from pathlib import Path

PO_FILE = Path("purchase_orders.csv")


def load_all() -> list[dict]:
    if not PO_FILE.exists():
        return []
    with PO_FILE.open(newline="", encoding="utf-8") as f:
        return [r for r in csv.DictReader(f)]


def find_match(po_number: str | None, vendor_name: str | None,
               amount: float | None) -> dict | None:
    """Find an open PO matching the invoice."""
    rows = [r for r in load_all() if (r.get("status") or "open").lower() == "open"]

    # Strategy 1: exact PO number match
    if po_number:
        po_clean = po_number.strip().upper()
        for r in rows:
            if (r.get("po_number") or "").strip().upper() == po_clean:
                return r

    # Strategy 2: vendor name + amount within $1 tolerance
    if vendor_name and amount is not None:
        v_clean = vendor_name.strip().lower()
        for r in rows:
            v_match = (r.get("vendor_name") or "").strip().lower() == v_clean
            try:
                a_match = abs(float(r.get("amount") or 0) - float(amount)) <= 1.0
            except Exception:
                a_match = False
            if v_match and a_match:
                return r

    return None
